### ISSUE TITLE
Real view matrices

### ISSUE THREAD

**Issue by DethRaid** [sp614x/optifine#2930]
Shader developers have been suffering. We have two uniform matrices, `gbufferModelView` and `gbufferModelViewInverse`, which are terrible

- They're not modelview matrices, they're just view matrices
- ...except that they don't have the camera's translation, only its rotation and scale, making transforming to/from view space more complicated than it needs to be

To transform a position from clip space to world space, we have to do
```glsl
vec4 world_pos = gbufferProjectionInverse * clip_pos;
world_pos = gbufferModelViewInverse * world_pos;
world_pos.xyz += cameraPosition;
```
There's no real reason why we have to add the camera position as a separate step, except that the matrix `gbufferModelViewInverse` (and it's sibling `gbufferModelView`) don't have the camera's translation

To solve both of these problems, I propose two new uniform matrices: `gbufferView` and `gbufferViewInverse`. These would be view matrices which encode the full transformation to/from view space. Shader developers won't need to suffer any more, and newcomers won't have to use a horribly named variable

**Comment by sp614x** [comment]
The shader devs shouldn't suffer.

**Comment by zombye** [comment]
No, shader developers have _not_ been "suffering" from this. Them not being named right is an incredibly minor issue, which while sure it should be fixed, should also be pretty low priority.

And NO, unless you want to make shader developers _actually_ suffer, don't include the full translation.

You do realize that the game lets you go 30 million blocks out, right?
This means you _can't_ get the full translation unless you want your highest spatial precision along X and/or Z to be measured in _blocks_ when out at that distance.

Even with cameraPosition you actually don't get the full translation, this was changed for the same reason.

There's also no reason whatsoever the view matrix should actually include the full translation, or even go to/from world-space at all, other than that this is what it normally does outside of Minecraft.

**Comment by DethRaid** [comment]
I definitely suffer

The precision issue isn't anything new. If you get the world space position today you'll run into the same problems - as you've pointed out. That's a limitation of the hardware we run on and it's not worth refusing to follow industry conventions

The reason is exactly that this is how things work in literally every 3d application _except_ for Minecraft. Minecraft isn't some special unique thing that needs highly specific rendering techniques. There's no actual reason why we can't add a real view matrix, except that it apparently I'm the only shader developer who isn't opposed to it

**Comment by DethRaid** [comment]
Also quick note that I don't want to get rid of any existing uniforms. People use them and apparently people like them and I'm not trying to take away anyone's workflow or make everyone rewrite their shaders. I just want to add two new uniforms

**Comment by zombye** [comment]
Blindly applying industry conventions to Minecraft without considering the differing needs of Minecraft and the rest of the industry is much, much worse than refusing to follow the industry conventions.

**Comment by zombye** [comment]
Have you actually looked at any 3D applications that are actually designed to render worlds at the scales Minecraft does, and what they are doing? Do _any_ of them use a "real" view matrix, except on the CPU side where they might be using one to calculate the other matrices they do use?

**Comment by zombye** [comment]
And is whatever you're doing actually going to benefit from a "real" view matrix, enough so that you "suffer" without one?
And are you "suffering" more than you would from the lack of precision when you _do_ get a "real" view matrix?

**Comment by Barteks2x** [comment]
Last time I looked into MC rendering code, it didn't even actually use the global world position for rendering. The positions used for rendering seem to wrap around every few thousand blocks, in some way (that details seems to be quite well hidden in the code, but it's there).

And just to show exactly what happens if you try to do it the trivial way, here is a recent video that shows it https://www.youtube.com/watch?v=cvHfrI2EUH4
You are going to get the same kind of issues in a shader that attempts to use global world coordinates.

**Comment by sp614x** [comment]
To avoid the rounding errors due to high coordinates the shaders use a hidden camera offset that gets updated above 1 million blocks or when the player teleports, rounded to 1000 blocks. The X/Z coordinates that the shaders see are not the real world coordinates.

**Comment by sp614x** [comment]
The matrices are dumped directly from OpenGL when rendering.
The camera is positioned at player feet and the chunks are rendered with eye offset added (this changed a little in one of the recent versions). The weirdness of the matrices is really a weirdness of the renderer itself.

**Comment by AntonioFerreras** [comment]
How would I go between view and world?
